junit-engine-api/src/main/java/org/junit/gen5/engine/AbstractTestDescriptor.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.gen5.engine;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public abstract class AbstractTestDescriptor implements TestDescriptor {
+
+	private final String uniqueId;
+	private TestDescriptor parent;
+	private final Set<TestDescriptor> children = new HashSet<>();
+
+	protected AbstractTestDescriptor(String uniqueId) {
+		this.uniqueId = uniqueId;
+	}
+
+	@Override
+	public boolean equals(Object other) {
+		if (other == null)
+			return false;
+		if (this.getClass() != other.getClass())
+			return false;
+		TestDescriptor otherDescriptor = (TestDescriptor) other;
+		return this.getUniqueId().equals(otherDescriptor.getUniqueId());
+	}
+
+	@Override
+	public int hashCode() {
```

Comment:
If hashing is expensive and the use case for adding the value to a set (or making it a key in a map) are unclear, then the Javadoc could explain that the class isn't designed to be a keg i a map.

In any case, glad we could a solution.